uint/src/uint.rs

```diff
@@ -321,15 +368,9 @@ macro_rules! impl_mulassign_from {
 
 #[inline(always)]
 #[doc(hidden)]
-pub fn mul_u32(a: (u64, u64), b: u64, carry: u64) -> (u64, u64) {
-	let upper = b * a.0;
-	let lower = b * a.1;
-
-	let (res1, overflow1) = lower.overflowing_add(upper << 32);
-	let (res2, overflow2) = res1.overflowing_add(carry);
-
-	let carry = (upper >> 32) + overflow1 as u64 + overflow2 as u64;
-	(res2, carry)
+pub fn mul_u64(a: u64, b: u64, carry: u64) -> (u64, u64) {
+	let (hi, lo) = split_u128(u128::from(a) * u128::from(b) + u128::from(carry));
```

Comment:
Is `u128` available on wasm?

Reply:
That's a good point. Seems to compile on [playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=069b9353269c926b95173d6fb0c65bd6), but I haven't measured perf for wasm. This method is needed for a faster division though.

Reply:
If you check, the playground actually inlines that function. You can check what it looks like under Wasm using [WebAssembly Studio](https://webassembly.studio/). It compiles, but if the function isn't inlined then it does a (relatively costly in Wasm) stack return instead of using 2 return values like one would expect:

```lisp
(func $split_u128 (export "split_u128") (type $t2) (param $p0 i32) (param $p1 i64) (param $p2 i64)
    get_local $p0
    get_local $p1
    i64.store offset=8
    get_local $p0
    get_local $p2
    i64.store)
```

So yes, this is available on Wasm, but I would be cautious. The assembly is actually fine, although laughably bad for how simple a function this is (it should just be 2x `mov r64, r64`), but assuming that you don't use the return value the next cycle after the call the cost will likely be hidden by the write being executed in parallel with other instructions:

```asm
wasm-function[1]:
  sub rsp, 8                            ; 0x000000 48 83 ec 08
  mov qword ptr [r15 + rdi + 8], rsi    ; 0x000004 49 89 74 3f 08
  mov qword ptr [r15 + rdi], rdx        ; 0x000009 49 89 14 3f
  nop                                   ; 0x00000d 66 90
  add rsp, 8                            ; 0x00000f 48 83 c4 08
  ret                                   ; 0x000013 c3
```

Reply:
Thanks Jack for the detailed info. This function is only used for parsing string (`from_dec_str`), multiplication by primitive types and (in #126) by a faster div. In div it's not a perf bottleneck, so I guess it's fine if it's currently not optimal for Wasm.